edgelet/edgelet-core/src/settings.rs

```diff
@@ -61,44 +112,113 @@ impl Manual {
-        )?;
+        let device_id =
+            device_id.ok_or(ErrorKind::ConnectionStringMalformedParameter(DEVICEID_KEY))?;
         let device_id_regex =
             Regex::new(DEVICEID_REGEX).expect("This hard-coded regex is expected to be valid.");
         if !device_id_regex.is_match(&device_id) {
-            return Err(ParseManualDeviceConnectionStringError::MalformedParameter(
+            return Err(Error::from(ErrorKind::ConnectionStringMalformedParameter(
                 DEVICEID_KEY,
-            ));
+            )));
         }
 
-        let hub = hub.ok_or(
-            ParseManualDeviceConnectionStringError::MissingRequiredParameter(HOSTNAME_KEY),
-        )?;
+        let hub = hub.ok_or(ErrorKind::ConnectionStringMissingRequiredParameter(
+            HOSTNAME_KEY,
+        ))?;
         let hub_regex =
             Regex::new(HOSTNAME_REGEX).expect("This hard-coded regex is expected to be valid.");
         if !hub_regex.is_match(&hub) {
-            return Err(ParseManualDeviceConnectionStringError::MalformedParameter(
+            return Err(Error::from(ErrorKind::ConnectionStringMalformedParameter(
                 HOSTNAME_KEY,
-            ));
+            )));
         }
 
         Ok((key, device_id.to_owned(), hub.to_owned()))
     }
 }
 
+#[derive(Clone, Debug, serde_derive::Deserialize, serde_derive::Serialize)]
+#[serde(tag = "method")]
+#[serde(rename_all = "lowercase")]
+pub enum ManualAuthMethod {
+    #[serde(rename = "device_connection_string")]
+    DeviceConnectionString(ManualDeviceConnectionString),
+    X509(ManualX509Auth),
+}
+
+#[derive(Clone, Debug, serde_derive::Serialize)]
+#[serde(rename_all = "lowercase")]
+pub struct Manual {
+    authentication: ManualAuthMethod,
+}
+
+impl<'de> serde::Deserialize<'de> for Manual {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        #[derive(Debug, serde_derive::Deserialize)]
+        struct Inner {
+            #[serde(skip_serializing_if = "Option::is_none")]
+            device_connection_string: Option<String>,
+            #[serde(skip_serializing_if = "Option::is_none")]
+            authentication: Option<ManualAuthMethod>,
+        }
+
+        let value: Inner = serde::Deserialize::deserialize(deserializer)?;
+
+        let authentication = match (value.device_connection_string, value.authentication) {
+            (Some(cs), Some(auth)) => {
+                if cs == DEFAULT_CONNECTION_STRING {
+                    auth
+                } else if let ManualAuthMethod::DeviceConnectionString(_) = auth {
+                    return Err(serde::de::Error::custom(
+                            "Multiple connection strings specified under provisioning.device_connection_string and provisioning.authentication.device_connection_string. Please specify only one connection string in the config.yaml",
```

Comment:
It doesn't really matter whether the `ManualAuthMethod` is a `DeviceConnectionString` or an `X509`. It should conflict with `device_connection_string` in either case. So don't check for what kind of `ManualAuthMethod` it is, and just return:

`"Only one of provisioning.device_connection_string and provisioning.authentication must be set in the config.yaml."`